File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos38/Alg38.cpp
/*
♦ Algoritmo 38: Algoritmo de implementação da Estrutura de Dados Não Linear Dinâmica (não Estática) de Grafo: Dijkstra.
*/


#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <limits>

template<typename T>
class Aresta {
private:
    No<T>* destino;
    int peso;

public:
    Aresta(No<T>* destino, int peso) : destino(destino), peso(peso) {}

    No<T>* getDestino() {
        return destino;
    }

    int getPeso() {
        return peso;
    }
};

template<typename T>
class No {
private:
    T valor;
    std::vector<Aresta<T>> arestas;

public:
    No(T valor) : valor(valor) {}

    T getValor() {
        return valor;
    }

    std::vector<Aresta<T>>& getArestas() {
        return arestas;
    }

    void adicionarAresta(No<T>* destino, int peso) {
        arestas.push_back(Aresta<T>(destino, peso));
    }
};

template<typename T>
class Grafo {
private:
    std::unordered_map<T, No<T>*> nos;

public:
    Grafo() {}

    void adicionarNo(T valor) {
        if (nos.find(valor) == nos.end()) {
            nos[valor] = new No<T>(valor);
        }
    }

    void adicionarAresta(T origem, T destino, int peso) {
        adicionarNo(origem);
        adicionarNo(destino);
        nos[origem]->adicionarAresta(nos[destino], peso);
    }

    std::unordered_map<T, int> dijkstra(T origem) {
        std::unordered_map<T, int> distancias;
        std::priority_queue<No<T>*, std::vector<No<T>*>, ComparadorDistancia<T>> filaPrioridade;

        for (const auto& par : nos) {
            distancias[par.first] = std::numeric_limits<int>::max();
        }

        distancias[origem] = 0;
        filaPrioridade.push(nos[origem]);

        while (!filaPrioridade.empty()) {
            No<T>* noAtual = filaPrioridade.top();
            filaPrioridade.pop();
            int distanciaAtual = distancias[noAtual->getValor()];

            for (const auto& aresta : noAtual->getArestas()) {
                No<T>* vizinho = aresta.getDestino();
                int novaDistancia = distanciaAtual + aresta.getPeso();

                if (novaDistancia < distancias[vizinho->getValor()]) {
                    distancias[vizinho->getValor()] = novaDistancia;
                    filaPrioridade.push(vizinho);
                }
            }
        }

        return distancias;
    }
};

int main() {
    Grafo<char> grafo;

    grafo.adicionarAresta('A', 'B', 10);
    grafo.adicionarAresta('A', 'C', 15);
    grafo.adicionarAresta('B', 'D', 12);
    grafo.adicionarAresta('B', 'F', 15);
    grafo.adicionarAresta('C', 'E', 10);
    grafo.adicionarAresta('D', 'E', 2);
    grafo.adicionarAresta('D', 'F', 1);
    grafo.adicionarAresta('F', 'E', 5);

    auto distancias = grafo.dijkstra('A');
    for (const auto& par : distancias) {
        std::cout << "Distância de A até " << par.first << ": " << par.second << std::endl;
    }

    return 0;
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos38/Alg38.rb
=begin
♦ Algoritmo 38: Algoritmo de implementação da Estrutura de Dados Não Linear Dinâmica (não Estática) de Grafo: Dijkstra.
=end


require 'set'
require 'priority_queue'

class Aresta
  attr_reader :destino, :peso
  
  def initialize(destino, peso)
    @destino = destino
    @peso = peso
  end
end

class No
  attr_reader :valor, :arestas
  
  def initialize(valor)
    @valor = valor
    @arestas = []
  end
  
  def adicionar_aresta(destino, peso)
    @arestas << Aresta.new(destino, peso)
  end
end

class Grafo
  def initialize
    @nos = {}
  end
  
  def adicionar_no(valor)
    @nos[valor] = No.new(valor) unless @nos.key?(valor)
  end
  
  def adicionar_aresta(origem, destino, peso)
    adicionar_no(origem)
    adicionar_no(destino)
    @nos[origem].adicionar_aresta(@nos[destino], peso)
  end
  
  def dijkstra(origem)
    distancias = Hash.new(Float::INFINITY)
    fila_prioridade = Containers::PriorityQueue.new { |a, b| distancias[a.valor] < distancias[b.valor] }
    
    @nos.keys.each { |valor| distancias[valor] = Float::INFINITY }
    
    distancias[origem] = 0
    fila_prioridade.push(@nos[origem])
    
    until fila_prioridade.empty?
      no_atual = fila_prioridade.pop
      distancia_atual = distancias[no_atual.valor]
      
      no_atual.arestas.each do |aresta|
        vizinho = aresta.destino
        nova_distancia = distancia_atual + aresta.peso
        
        if nova_distancia < distancias[vizinho.valor]
          distancias[vizinho.valor] = nova_distancia
          fila_prioridade.push(vizinho)
        end
      end
    end
    
    distancias
  end
end

grafo = Grafo.new

grafo.adicionar_aresta('A', 'B', 10)
grafo.adicionar_aresta('A', 'C', 15)
grafo.adicionar_aresta('B', 'D', 12)
grafo.adicionar_aresta('B', 'F', 15)
grafo.adicionar_aresta('C', 'E', 10)
grafo.adicionar_aresta('D', 'E', 2)
grafo.adicionar_aresta('D', 'F', 1)
grafo.adicionar_aresta('F', 'E', 5)

distancias = grafo.dijkstra('A')
distancias.each do |vertice, distancia|
  puts "Distância de A até #{vertice}: #{distancia}"
end
